desc/protoparse/ast/file.go

```diff
@@ -0,0 +1,234 @@
+// protobuf source file.
+type FileNode struct {
+	compositeNode
+	Syntax *SyntaxNode // nil if file has no syntax declaration
+	Decls  []FileElement
+
+	// Any comments that follow the last token in the file.
+	FinalComments []Comment
+	// Any whitespace at the end of the file (after the last token or
+	// last comment in the file).
+	FinalWhitespace string
+}
+
+// NewFileElement creates a new *FileNode. The syntax parameter is optional. If it
+// is absent, it means the file had no syntax declaration.
+//
+// This function panics if the concrete type of any element of decls is not
+// from this package.
+func NewFileNode(syntax *SyntaxNode, decls []FileElement) *FileNode {
+	numChildren := len(decls)
+	if syntax != nil {
+		numChildren++
+	}
+	children := make([]Node, 0, numChildren)
+	if syntax != nil {
+		children = append(children, syntax)
+	}
+	for _, decl := range decls {
+		children = append(children, decl)
+	}
+
+	for _, decl := range decls {
+		switch decl := decl.(type) {
+		case *PackageNode, *ImportNode, *OptionNode, *MessageNode,
+			*EnumNode, *ExtendNode, *ServiceNode, *EmptyDeclNode:
+		default:
+			panic(fmt.Sprintf("invalid FileElement type: %T", decl))
+		}
+	}
+
+	return &FileNode{
+		compositeNode: compositeNode{
+			children: children,
+		},
+		Syntax: syntax,
+		Decls:  decls,
+	}
+}
+
+func NewEmptyFileNode(filename string) *FileNode {
+	return &FileNode{
+		compositeNode: compositeNode{
+			children: []Node{NewNoSourceNode(filename)},
+		},
+	}
+}
+
+func (f *FileNode) GetSyntax() Node {
+	return f.Syntax
+}
+
+// FileElement is an interface implemented by all AST nodes that are
+// allowed as top-level declarations in the file.
+type FileElement interface {
+	Node
+	fileElement()
+}
+
+var _ FileElement = (*ImportNode)(nil)
+var _ FileElement = (*PackageNode)(nil)
+var _ FileElement = (*OptionNode)(nil)
+var _ FileElement = (*MessageNode)(nil)
+var _ FileElement = (*EnumNode)(nil)
+var _ FileElement = (*ExtendNode)(nil)
+var _ FileElement = (*ServiceNode)(nil)
+var _ FileElement = (*EmptyDeclNode)(nil)
+
+// SyntaxNode represents a syntax declaration, which if present must be
+// the first non-comment content. Example:
+//
+//  syntax = "proto2";
+//
+// Files that don't have a syntax node are assumed to use proto2 syntax.
+type SyntaxNode struct {
+	compositeNode
+	Keyword   *KeywordNode
+	Equals    *RuneNode
+	Syntax    StringValueNode
+	Semicolon *RuneNode
+}
+
+// NewSyntaxNode creates a new *SyntaxNode. All four arguments must be non-nil:
+//  - keyword: The token corresponding to the "syntax" keyword.
+//  - equals: The token corresponding to the "=" rune.
+//  - syntax: The actual syntax value, e.g. "proto2" or "proto3".
+//  - semicolon: The token corresponding to the ";" rune that ends the declaration.
+func NewSyntaxNode(keyword *KeywordNode, equals *RuneNode, syntax StringValueNode, semicolon *RuneNode) *SyntaxNode {
+	if keyword == nil {
+		panic("keyword is nil")
```

Comment:
Just making sure - none of these are triggered if I input a file that has a missing keyword/equals/etc, this is just saying "if these are hit, the code inside protoparse/ast was constructed incorrectly - right?

Reply:
Correct. The grammar requires all of the tokens be in the right place and will report a syntax error otherwise instead of trying to create a junk AST node.

Reply:
For example, the parser would not even know this could be a `SyntaxNode` (so it wouldn't even know to call `NewSystaxNode`) if there is no `syntax` keyword in the source file.

Reply:
And if you're still worried, I encourage applying the fuzz tester to this branch (if that is easy to do) to verify there are no panics. If there are, they obviously need to be remedied before merging.

Reply:
Unfortunately `go-fuzz` is broken with 1.15 so fuzzing is a little more annoying right now :grin:.